lib/kube/token/validator.go

```diff
@@ -85,35 +83,59 @@ func (c *ValidationResult) JoinAuditAttributes() (map[string]interface{}, error)
 // Kubernetes TokenRequest API endpoint.
 type TokenReviewValidator struct {
 	mu sync.Mutex
-	// client is protected by mu and should only be accessed via the getClient
-	// method.
-	client kubernetes.Interface
+	// client and clusterAudiences are protected by mu and should only be
+	// accessed via the getClient method.
+	client           kubernetes.Interface
+	clusterAudiences []string
 }
 
 // getClient allows the lazy initialisation of the Kubernetes client
-func (v *TokenReviewValidator) getClient() (kubernetes.Interface, error) {
+func (v *TokenReviewValidator) getClient(ctx context.Context) (kubernetes.Interface, []string, error) {
 	v.mu.Lock()
 	defer v.mu.Unlock()
-	if v.client != nil {
-		return v.client, nil
+	if v.client != nil && v.clusterAudiences != nil {
+		return v.client, v.clusterAudiences, nil
 	}
 
 	config, err := rest.InClusterConfig()
 	if err != nil {
-		return nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes config")
+		return nil, nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes config")
 	}
 	client, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		return nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes client")
+		return nil, nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes client")
+	}
+
+	// We do a self-review to recover the default cluster audience
+	audiences, err := getTokenAudiences(ctx, client, config.BearerToken)
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "doing a self-review")
 	}
 
 	v.client = client
-	return client, nil
+	v.clusterAudiences = audiences
+	return client, audiences, nil
+}
+
+func getTokenAudiences(ctx context.Context, client kubernetes.Interface, token string) ([]string, error) {
+	review := &v1.TokenReview{
+		Spec: v1.TokenReviewSpec{
+			Token: token,
+		},
+	}
+	options := metav1.CreateOptions{}
+
+	reviewResult, err := client.AuthenticationV1().TokenReviews().Create(ctx, review, options)
+	if err != nil {
+		return nil, trace.Wrap(err, "reviewing token and retrieving audience")
+	}
+
+	return reviewResult.Status.Audiences, nil
```

Comment:
The [API docs](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-review-v1/#TokenReviewStatus) for `status.audiences` mention 

>  a TokenReview returns an empty status.audience field where status.authenticated is "true", the token is valid against the audience of the Kubernetes API server.

Can we be reasonably sure that all supported kube api servers will actually populate the field?

Reply:
If Kubernetes  doesn't guarantee that the `audiences` field is always defined, we should prioritize parsing the JWT token directly, as this ensures we know the field is set. My concern is that, since the documentation doesn't mandate it, there might have been a version in the past that didn't include the `audiences` field, or a future version might omit it. To safeguard against this, we should handle the token by unmarshaling it ourselves.


You can use jwt/v4 package for that

```go
import "github.com/golang-jwt/jwt/v4"
...
p := jwt.NewParser()
	claims := &jwt.RegisteredClaims{}
	_, _, err := p.ParseUnverified(token, claims)
	if err != nil {
		panic(err.Error())
	}

	fmt.Println(claims.Audiences)
}



```

Reply:
I changed to parsing our own token without validating. The missing audience thing is for legacy Kubernetes tokens (which are in theory still supported, but we do reject them for kube >= 1.21).

So basically:
- if there's an audience on our token we know this is a modern kube and can set the audience
- if there are no audience on our token, this is an old kube and we do the review without specifying audience, as we were doing before